tests for cli, #1387

bylsmad opened the pull request:
implementing tests for:
help
create-instance
index-create
load-help
item-get
item-put
index-dump

removed newline at end of file for *.data files
caused failure in item-get; item-put test

TODO: add tests for encoding issues, see #1378
TODO: create tests for other commands

bylsmad commented on src/moin/cli/_tests/test_modify_item.py:
```diff
@@ -0,0 +1,100 @@
+    assert cat_bytes_expected == cat_bytes
+
+
+def test_item_put(index_create):
```
this one really should be using index_create2 to insure that it doesn't mess with test_dump_help

bylsmad commented on _test_artifacts/README.rst:
```diff
@@ -0,0 +1,6 @@
+
+this directory contains artifacts from tests
+
+any files dropped here will be uploaded in github action
```
looks like I missed the change to ci.yml which makes this statement true, since these tests don't leave anything behind in _test_artifacts it is not needed, will make the ci.yml change in another branch where there are actual files left behind to archive

UlrichB22 commented:
Thanks for adding tests to cli. Some thoughts about the code:

- new python modules should get a header with copyright according to [src/moin/_template.py](https://github.com/moinwiki/moin/blob/master/src/moin/_template.py). Instead of `MoinMoin:YourNameHere` I commonly use `MoinMoin project` 
- Is it possible to keep the *.data files as they are? IMO every text file should have a linefeed at the end. What was the problem?
- Can we avoid having a new directory _test_artifacts in the source code as it seems to be created anyway when it is missing? This would help to keep the code simple and clean.
- I want to remove `-i -s` options on `moin index-create`, they are already obsolete. Can you try without?

I didn't test yet, this will need some time.

bylsmad commented:
the issue with the ending newlines was an item in storage which does not have a trailing newline would raise this exception if you item-get then item-put the resulting files:

`ValueError: computed data size (18) does not match data size declared in metadata (16)`

this issue is now resolved and the ending newlines are added back to the help/en files

I've addressed all the other issues you raised in your review, except I've left the non-english item in the help-en directory as it is the cause of the test failure in #1378 ... will move it in that branch

RogerHaase commented:
I am getting 4 file not found errors running tests on Windows.

https://gist.github.com/RogerHaase/921008302d20d30eec2764811ee15023

bylsmad commented:
@RogerHaase please retest, fix for windows pushed https://github.com/moinwiki/moin/pull/1389/commits/9351fa6c227fca526b6afc62ac4a11e05353533d

bylsmad commented:
@RogerHaase there will be a merge conflict once #1391 is merged ... we both fixed the issue with the trailing line endings created by  item-get in slightly different ways...I have a rebased version of my branch ready to go with fixes for the test, so let me know once #1391 is merged and I will push my rebased branch

UlrichB22 commented:
Please also merge changes to index.py and modify_item.py from #1388. Thanks.

bylsmad commented:
@RogerHaase update: I investigated the line endings in the repo by setting my gitconfig `core.autocrlf false` and cloning your fork.  I found that even in the help branch both the help .data files as well as the .py files are stored in git with \n.  I've added comment on #1391 I am going to hold off on making the line ending changes in this branch

@UlrichB22 I just rebased this branch to the latest master which includes #1392 and #1388 cli tests are still passing

RogerHaase commented:
Running tests on windows now down to 1 error

https://gist.github.com/RogerHaase/547f111e971cc6ba5fe73824c5aef0d2

bylsmad commented:
@RogerHaase you found two bugs, one in the test one in cli  see https://github.com/moinwiki/moin/pull/1389/commits/50e7bacf3de46a69ceb3fc2e1a1b0833d94533d8

RogerHaase commented:
Thanks for helping!
